### PR TITLE
Jj/feature/live rec

### PR DESCRIPTION
**Pull request by jjauzion**
New feature:  
- Live recording picture (with predicted label) while the patate is running in autonomous mode (debug mode = 1).
- Measure of the preditiction rate (nb of prediction per second)

Only 1 picture over 4 is saved
The picture are saved in a buffer that containing 100 pictures max. The buffer is written to the disk when the car is stopped.

No performance penalty observed in debug mode

### CODE REVIEW

File: race.py [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
 
 import Adafruit_PCA9685
 # noinspection PyUnresolvedReferences
 from keras.models import load_model
 
-from pivideostream import PiVideoStream
+from utils.pivideostream import PiVideoStream
 from utils.const import SPEED_NORMAL, SPEED_FAST, HEAD_UP, HEAD_DOWN, \
     DIRECTION_R, DIRECTION_L, DIRECTION_C, DIRECTION_L_M, DIRECTION_R_M
+from utils.path import OUTPUT_DIRECTORY
```
Comment (blopax) [inline race.py]: Faudra qu on s accorde sur les output path il y a des cas differents (test, get_data, race, training et validation pour le train). C'est pas urgent ca dependra un peu de ce que tu fais ac S3.

File: race.py [hunk 2]
```diff
@@ -113,7 +149,7 @@ def stop(self):
         # TODO (pclement): other inputs to stop the motor & direct the wheels without having to reload the full model.
 
         starting_prompt = """Press 'go' + enter to start.
-        Press 'show' + enter to start with the printing mode.
+        Press 'debug=$LEVEL_VAL' + enter to start in debug mode of level LEVEL_VAL (ex: debug=1)
```
Comment (blopax) [inline race.py]: Faudrait sans doute decrire les differents niveaux

File: race.py [hunk 3]
```diff
@@ -67,31 +75,45 @@ def choose_head(predictions, speed):
...
 
             # Get model prediction
             predictions_raw = self.model.predict(image)
             predictions = [np.argmax(pred, axis=1) for pred in predictions_raw]  # Why ?
 
             # Decide action
             direction = self.choose_direction(predictions)
-            head = self.choose_head(predictions, speed)
             speed = self.choose_speed(predictions)
+            head = self.choose_head(predictions, speed)
```
Comment (blopax) [inline race.py]: pourquoi tu le rechanges ? c'est toi qui me l'avait fait changer :)

File: race.py [hunk 4]
```diff
@@ -1,20 +1,25 @@
...
 from utils.const import SPEED_NORMAL, SPEED_FAST, HEAD_UP, HEAD_DOWN, \
     DIRECTION_R, DIRECTION_L, DIRECTION_C, DIRECTION_L_M, DIRECTION_R_M
+from utils.path import OUTPUT_DIRECTORY
 
 
 def get_args():
     parser = argparse.ArgumentParser()
     parser.add_argument("model_path", type=str,
                         help="Provide the model path.\n")
+    parser.add_argument("-t", "--trial", action="store_true",
```
Comment (blopax) [inline race.py]: ca fait pas un doublon avec le mode debug=1 ? Pourquoi demander les 2 ?

File: race.py [hunk 5]
```diff
@@ -100,6 +122,20 @@ def stop(self):
...
+        # Performance status
+        elapse_time = time.time() - self.start_time
+        pred_rate = self.nb_pred / float(elapse_time)
+        print('{} prediction in {}s -> {} pred/s'.format(self.nb_pred, elapse_time, pred_rate))
+
+        # Write buffer
+        if self.buffer is not None and len(self.buffer) > 0:
+            print('Saving buffer pictures to : "{}"'.format(OUTPUT_DIRECTORY))
+            for i, img in enumerate(self.buffer):
+                pic_file = '{}_image{}_{}_{}.jpg'.format(self.start_time, i, img["speed"], img["direction"])
```
Comment (blopax) [inline race.py]: On pourrait creuser le fait de mettre le timestamp. Mais effectivement ca peut avoir un impact sur les perfs.

**Comment by blopax** [comment]
Qq petites remarques mais au vu des tests ca me semble bien. Juste un point a verifier avant que j'approuve: cette histoire de choose.head() et choose.direction(). On peut en parler sur slack !

File: race.py (hunk 1 above)
Comment (jjauzion) [reply race.py]: Oui en effet!

File: race.py (hunk 2 above)
Comment (jjauzion) [reply race.py]: Oui, je voulais pas trop alourdir le print mais au moins mettre les valeurs acceptés serait bien

File: race.py (hunk 3 above)
Comment (jjauzion) [reply race.py]: hum étrange, c'est pas volontaire, j'ai du fucked up quand j'ai fais le rebase. Moralité: penser a vérifier les changes soit même après une PR avant de la soumettre a revue ^^

File: race.py (hunk 4 above)
Comment (jjauzion) [reply race.py]: Oui bien vu! J'avais ça avant le mode debug et j'ai pas pensé a l'enlever après le rebase

File: race.py (hunk 5 above)
Comment (jjauzion) [reply race.py]: Vu que la on est dans le stop (ie, la voiture est arrêté), ça n'impacte pas les perf normalement, et ça permet de ne pas écraser les images sauver du précédent ctrl+c

File: race.py [hunk 6]
```diff
@@ -61,19 +67,22 @@ def choose_speed(predictions):
             return SPEED_FAST
         return SPEED_NORMAL
 
-    @staticmethod
-    def choose_head(predictions, speed):
-        if speed == SPEED_FAST and predictions[1] == 1 and predictions[0] == 1:
+    def choose_head(self, predictions, speed):
+        if speed == self.choose_speed(predictions) == SPEED_FAST:
```
Comment (blopax) [inline race.py]: nice!

File: race.py [hunk 7]
```diff
@@ -61,19 +67,22 @@ def choose_speed(predictions):
             return SPEED_FAST
```
Comment (blopax) [inline race.py]: La du coup il faudra changer et remplacer par les constantes... c'est faux de mettre 1 maintenant qu on est a 5 directions. (ca ne change pas la PR)
